content/docs/userspace/full-stack/4-json.md

```diff
@@ -0,0 +1,355 @@
+
+The mark conversion will be done by the corresponding mark file in `/mar` on the
+agent's desk. In our case it would be `/mar/journal/action.hoon` and
+`/mar/journal/update.hoon` in the `%journal` desk for our `%journal-action` and
+`%journal-update` marks, which are for the `$action` and `$update` structures we
+defined previously.
+
+Mark conversion functions can be included directly in the mark file, or they can
+be written in a separate library, then imported and called by the mark file. We
+will do the latter in this case, so before we create the mark files themselves,
+we'll write a library called `/lib/journal.hoon` with the conversion functions.
+
+## `$json` utilities
+
+[`zuse.hoon`](/docs/hoon/reference/zuse/table-of-contents) contains three main
+cores for converting to and from `$json`:
+
+- [`++enjs:format`](/docs/hoon/reference/zuse/2d_1-5#enjsformat) - Functions to
+  help encode data structures as `$json`.
+- [`++dejs:format`](/docs/hoon/reference/zuse/2d_6#dejsformat) - Functions to
+  decode `$json` to other data structures.
+- [`++dejs-soft:format`](/docs/hoon/reference/zuse/2d_7#dejs-softformat) -
+  Mostly the same as `++dejs:format` except the functions produce units which
+  are null if decoding fails, rather than just crashing.
+
+### `++enjs:format`
+
+This contains ten functions for encoding `$json`. Most of them are for specific
+hoon data types, such as `++tape:enjs:format`, `++ship:enjs:format`,
+`++path:enjs:format`, etc. We'll just have a look at the two most general and
+useful ones: `++frond:enjs:format` and `++pairs:enjs:format`.
+
+#### `++frond`
+
+This function is for forming a JSON object from a single key-value pair. For
+example:
+
+```
+> (frond:enjs:format 'foo' s+'bar')
+[%o p={[p='foo' q=[%s p='bar']]}]
+```
+
+When stringified by Eyre, this will look like:
+
+```json
+{ "foo": "bar" }
+```
+
+#### `++pairs`
+
+This is similar to `++frond` and also forms a JSON object, but it takes multiple
+key-value pairs rather than just one:
+
+```
+> (pairs:enjs:format ~[['foo' n+~.123] ['bar' s+'abc'] ['baz' b+&]])
+[%o p={[p='bar' q=[%s p='abc']] [p='baz' q=[%b p=%.y]] [p='foo' q=[%n p=~.123]]}]
+```
+
+When stringified by Eyre, this will look like:
+
+```json
+{
+  "foo": 123,
+  "baz": true,
+  "bar": "abc"
+}
+```
+
+Notice that we used a knot for the value of `foo` (`n+~.123`). Numbers in JSON
+can be signed or unsigned and integers or floating point values. The `$json`
+structure uses a knot so that you can decide whether a particular number should
+be treated as `@ud`, `@sd`, `@rs`, etc.
+
+### `++dejs:format`
+
+This core contains many functions for decoding `$json`. We'll touch on some
+useful families of `++dejs` functions in brief, but because there's so many, in
+practice you'll need to look through the [`++dejs`
+reference](/docs/hoon/reference/zuse/2d_6) to find the correct functions for
+your use case.
+
+#### Number functions
+
+- `++ne` - decode a number to a `@rd`.
```

Comment:
any mnemonic for these?

e = engineering?
i = int
o = object?

Reply:
yeah presumably `i` is for integer, `o` for object etc. I don't think there's an overall system to it though & not all of them have a clear mapping. I don't think it's useful for people to try and memorise them all, easier to just look them up as you need them